src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -1305,182 +1305,224 @@ protected override void HandlePauseRequestAsync(IRequestResponder<PauseArguments
+                            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+                            try
                             {
-                                if (m_moduleMap.TryGetValue(module, out int mapModuleId))
+                                CancellationToken cancellationToken = cancellationTokenSource.Token;
+                                Task[] frameTasks = response.StackFrames.Select(frame => Task.Run(() =>
+                                {
+                                    cancellationToken.ThrowIfCancellationRequested();
+                                    if (frame.Source != null && !File.Exists(frame.Source.Path))
+                                    {
+                                        cancellationToken.ThrowIfCancellationRequested();
+                                        // mark the stack frame to be skipped by default
+                                        frame.Source.PresentationHint = Source.PresentationHintValue.Deemphasize;
+                                    }
+                                })).ToArray();
+
+                                Task allFramesComplete = Task.WhenAll(frameTasks);
+                                Task timeoutTask = Task.Delay(20000, cancellationToken);
+                                try
+                                {
+                                    await Task.WhenAny(allFramesComplete, timeoutTask);
+                                }
+                                catch (Exception e)
                                 {
-                                    moduleId = mapModuleId;
+                                    Debug.Fail("Ignoring exception computing frame state: " + e.ToString());
                                 }
                             }
+                            finally
+                            {
+                                cancellationTokenSource.Cancel();
+                                cancellationTokenSource.Dispose();
+                            }
                         }
 
-                        response.StackFrames.Add(new ProtocolMessages.StackFrame()
-                        {
-                            Id = frameReference,
-                            Name = frameInfo.m_bstrFuncName,
-                            Source = textPosition.Source,
-                            Line = textPosition.Line,
-                            Column = textPosition.Column,
-                            ModuleId = moduleId
-                        });
+                        response.TotalFrames = (int)frameEnumInfo.TotalFrames;
                     }
-
-                    response.TotalFrames = (int)frameEnumInfo.TotalFrames;
                 }
-            }
 
-            responder.SetResponse(response);
+                responder.SetResponse(response);
+            }
+            catch (Exception e)
+            {
+                responder.SetError(new ProtocolException(e.ToString()));
```

Comment:
>.ToString() [](start = 58, length = 11)

`ToString()` will include the stack trace of the thrown exception. Do you want `Message` instead?